provisioning/applysecret.go

```diff
@@ -0,0 +1,104 @@
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coreclientv1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+)
+
+type updateDataFn func(existing, required *corev1.Secret) error
+
+func doNotUpdateData(existing, required *corev1.Secret) error {
+	return nil
+}
+
+// applySecret merges objectmeta, applies data if the secret does not exist.
+// Note: this is mostly a copy of resourceapply.ApplySecret() except we allow to choose what
+//       data to update. This is because we don't want to keep rotating passwords.
+func applySecret(client coreclientv1.SecretsGetter, requiredInput *corev1.Secret, updateData updateDataFn) error {
+	// copy the stringData to data.  Error on a data content conflict inside required.  This is usually a bug.
+	required := requiredInput.DeepCopy()
+	if required.Data == nil {
+		required.Data = map[string][]byte{}
+	}
+	for k, v := range required.StringData {
+		if dataV, ok := required.Data[k]; ok {
+			if string(dataV) != v {
+				return fmt.Errorf("Secret.stringData[%q] conflicts with Secret.data[%q]", k, k)
+			}
+		}
+		required.Data[k] = []byte(v)
+	}
+	required.StringData = nil
+
+	existing, err := client.Secrets(required.Namespace).Get(context.TODO(), required.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		requiredCopy := required.DeepCopy()
+		_, err := client.Secrets(requiredCopy.Namespace).
+			Create(context.TODO(), resourcemerge.WithCleanLabelsAndAnnotations(requiredCopy).(*corev1.Secret), metav1.CreateOptions{})
+		return err
+	}
+	if err != nil {
+		return err
+	}
+
+	existingCopy := existing.DeepCopy()
+
+	resourcemerge.EnsureObjectMeta(resourcemerge.BoolPtr(false), &existingCopy.ObjectMeta, required.ObjectMeta)
+
+	// Allow the caller to decide what data to update.
+	err = updateData(existingCopy, required)
+	if err != nil {
+		return err
+	}
+
+	existingCopy.Type = required.Type
+
+	// Server defaults some values and we need to do it as well or it will never equal.
+	if existingCopy.Type == "" {
+		existingCopy.Type = corev1.SecretTypeOpaque
+	}
+
+	if equality.Semantic.DeepEqual(existingCopy, existing) {
+		return nil
+	}
+
+	if klog.V(4).Enabled() {
+		klog.Infof("Secret %s/%s changes: %v", required.Namespace, required.Name, resourceapply.JSONPatchSecretNoError(existing, existingCopy))
+	}
+
+	/*
+	 * Kubernetes validation silently hides failures to update secret type.
+	 * https://github.com/kubernetes/kubernetes/blob/98e65951dccfd40d3b4f31949c2ab8df5912d93e/pkg/apis/core/validation/validation.go#L5048
+	 * We need to explicitly opt for delete+create in that case.
+	 */
```

Comment:
Since we are essentially writing our own applySecret() should we take care of a scenario where the secret.Type is changing? AFAIK, we do not expect any of our Secrets to change their type. So, in that case, should we just be returning an error instead of deleting and creating the Secret?

Reply:
Ok. I see thet existingCopy.Type has been updated to required.Type or corev1.SecretTypeOpaque.

Reply:
not sure, i thought it would be easier to compare if they are mostly the same?